mlem/contrib/venv.py

```diff
@@ -0,0 +1,228 @@
+
+
+class EnvBuilder(MlemBuilder):
+    """MlemBuilder implementation for building virtual environments"""
+
+    type: ClassVar = "env"
+
+    target: Optional[str] = "venv"
+    """Name of the virtual environment"""
+
+    @abstractmethod
+    def create_virtual_env(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_installed_packages(self, env_dir):
+        raise NotImplementedError
+
+    @abstractmethod
+    def build(self, obj: MlemModel) -> str:
+        raise NotImplementedError
+
+
+class VenvBuilder(EnvBuilder):
+    class Config:
+        arbitrary_types_allowed = True
+
+    type: ClassVar = "venv"
+
+    no_cache: Optional[bool] = False
+    """Disable cache"""
+    current_env: Optional[bool] = False
+    """Whether to install in the current virtual env, must be active"""
+    context: Optional[SimpleNamespace] = None
+    """context for the virtual env"""
+
+    def create_virtual_env(self):
+        env_spec = venv.EnvBuilder(with_pip=True)
+        env_dir = os.path.abspath(self.target)
+
+        self.context = env_spec.ensure_directories(env_dir)
+
+        true_system_site_packages = env_spec.system_site_packages
+        env_spec.system_site_packages = False
+        env_spec.create_configuration(self.context)
+        env_spec.setup_python(self.context)
+        if env_spec.with_pip:
+            env_spec._setup_pip(  # pylint: disable=protected-access
+                self.context
+            )
+        if not env_spec.upgrade:
+            env_spec.setup_scripts(self.context)
+            env_spec.post_setup(self.context)
+        if true_system_site_packages:
+            env_spec.system_site_packages = True
+            env_spec.create_configuration(self.context)
+
+    def get_installed_packages(self, env_dir):
+        if platform.system() == "Windows":
+            env_exe = os.path.join(env_dir, "Scripts", "python.exe")
+        else:
+            env_exe = os.path.join(env_dir, "bin", "python")
+
+        try:
+            return subprocess.check_output([env_exe, "-m", "pip", "freeze"])
+        except FileNotFoundError as e:
+            raise MlemError(f"Executable for pip not found.\n{e}") from e
+        except subprocess.CalledProcessError as e:
+            raise MlemError(
+                f"Couldn't get the list of packages from pip.\n{e}"
+            ) from e
+        except subprocess.TimeoutExpired as e:
+            raise MlemError(
+                f"Determining list of pip packages timed out.\n{e}"
+            ) from e
+
+    def build(self, obj: MlemModel):
+        if self.current_env:
+            if (
+                os.getenv("VIRTUAL_ENV") is None
+                or sys.prefix == sys.base_prefix
+            ):
+                raise MlemError("No virtual environment detected.")
+            echo(EMOJI_PACK + f"Detected the virtual env {sys.prefix}")
+            env_dir = sys.prefix
+            if platform.system() == "Windows":
```

Comment:
seems like a good candidate for a separate method

Reply:
done